core/src/main/scala/org/http4s/blaze/channel/ServerChannel.scala

```diff
@@ -1,67 +1,114 @@
+    // inside the lock since the EC could run the task immediately.
+    if (run) scheduleHook()
+  }
 
-    l.await()
+  /** Wait for this server channel to close, including execution of all successfully
+    * registered shutdown hooks.
+    */
+  final def join(): Unit = shutdownHooks.synchronized {
+    while (state != Closed) {
+      shutdownHooks.wait()
+    }
   }
 
   /** Add code to be executed when the [[ServerChannel]] is closed
     *
     * @param f hook to execute on shutdown
-    * @return true if the hook was successfully registered, false otherwise
+    * @return true if the hook was successfully registered, false otherwise.
     */
-  final def addShutdownHook(f: () => Unit): Boolean = {
-    @tailrec
-    def go(): Boolean = shutdownHooks.get() match {
-      case null                                                     => false
-      case hooks if(shutdownHooks.compareAndSet(hooks, hooks :+ f)) => true
-      case _                                                        => go()
+  final def addShutdownHook(f: () => Unit)(implicit ec: ExecutionContext = Execution.directec): Boolean = {
+    shutdownHooks.synchronized {
+      if (state != Open) false
+      else {
+        shutdownHooks += Hook(f, ec)
+        true
+      }
     }
+  }
 
-    go()
+  // Checks if we have a hook, and if we do, schedule it and then check again
+  private[this] def scheduleHook(): Unit = {
+    val hook = shutdownHooks.synchronized {
+      if (shutdownHooks.nonEmpty) Some(shutdownHooks.dequeue())
+      else {
+        // all our hooks are done! Switch to closed state, notify
+        // all the listeners
+        state = Closed
+        shutdownHooks.notifyAll()
+        None
+      }
+    }
+    hook match {
+      case Some(hook) => hook.ec.execute(new HookRunnable(hook.task))
+      case None => ()
+    }
   }
 
-  private def runShutdownHooks(): Unit = {
-    val hooks = shutdownHooks.getAndSet(null)
-    if (hooks != null) {
-      hooks.foreach { f =>
-        try f()
-        catch { case NonFatal(t) => logger.error(t)(s"Exception occurred during Channel shutdown.") }
+  private case class Hook(task: () => Unit, ec: ExecutionContext)
+
+  // Bundles the task of executing a hook and scheduling the next hook
+  private[this] class HookRunnable(hook: () => Unit) extends Runnable {
+    override def run(): Unit = {
+      try hook()
+      catch {
+        case NonFatal(t) =>
+          logger.error(t)(s"Exception occurred during Channel shutdown.")
+      }
+      finally {
+        // We bounce these through the thread local trampoline to ensure
+        // that we don't get a SOE if everyone is using the direct EC.
+        Execution.trampoline.execute(new Runnable {
+          override def run(): Unit = scheduleHook()
+        })
       }
     }
   }
+}
 
-  /* Return the bound socket address for this server channel */
-  def socketAddress: InetSocketAddress
+private object ServerChannel {
+  object State extends Enumeration {
```

Comment:
To avoid making 4 new classes/traits to clutter the classpath, which I suppose is a drop in the ocean for Scala. I'd definitely choose the sealed + case objects pattern of I did any matching, but it's all equality checks.
Is this a blocker for either of you? I'm not strongly opinionated about it, this just felt 'right' and is less code (by 1 line).